### ISSUE TITLE
Sending RELP messages bigger than maxMessageSize never gets acked and will seg fault when rsyslog is stopped

### ISSUE THREAD

**Issue by StrongestNumber9** [rsyslog/rsyslog#2633]
### Expected behavior
Receiver either acks and trims, or alternatively declines messages above maxMessageSize in a way client understands to stop sending them. Rsyslog shuts down correctly.

### Actual behavior
Both clients seg fault when stopped. Sender will keep on sending something that is rejected.

### Steps to reproduce the behavior
Prepare log
```sh
for i in {1..3}; do
    head -c 300000 /dev/zero | tr '\0' '=' >> /input.log;
    echo >> /input.log
done;
```

Sender configuration
```sh
$LocalHostName sender
$AbortOnUncleanConfig on
$PreserveFQDN on
$FileCreateMode 0644
global(
    workDirectory="/var-sender"
    maxMessageSize="128k"
)
module(load="omrelp")
module(load="imfile")
input(type="imfile" file="/input.log" ruleset="input" tag="1k")

ruleset(name="input" queue.type="direct"){
    action
    (
     type="omrelp"
     name="omrelp"
     target="127.0.0.1"
     port="601"
     queue.size="12345"
     queue.type="Disk"
     queue.checkpointinterval="65536"
     queue.filename="queue"
     queue.maxfilesize="128m"
     queue.saveonshutdown="on"
     queue.syncqueuefiles="off"
     queue.maxdiskspace="1g"
     action.ResumeInterval="2"
     action.ResumeRetryCount="-1"
     RebindInterval="1000"
  )
}
```

Receiver (on same machine)
```sh
$LocalHostName receiver
$AbortOnUncleanConfig on
$PreserveFQDN on
$FileCreateMode 0644
global(
    workDirectory="/var-receiver"
    maxMessageSize="128k"
)
module(load="imrelp")
input(type="imrelp" port="601" ruleset="input")
ruleset(name="input" queue.type="direct"){
    action(type="omfile" file="/output.log")
}
```

Running commands
```sh
# rsyslogd -f /rsyslog-receiver.conf -i /var-receiver/rsyslogd.pid -dn
# rsyslogd -f /rsyslog-sender.conf -i /var-sender/rsyslogd.pid -dn
```

Example of what RELP messages look like from receivers point of view
```sh
5797.375890297:imrelp.c       : imrelp.c: librelp: done epoll_wait, nEvents:1
5797.375908680:imrelp.c       : imrelp.c: new connect on RELP socket #5
5797.376037591:imrelp.c       : imrelp.c: remote host is 'localhost', ip '127.0.0.1'
5797.376050874:imrelp.c       : imrelp.c: librelp: add socket 8 to epoll set (ptr 0x7f72a8002400)
5797.376061360:imrelp.c       : imrelp.c: librelp: epoll_set_events sock 8, target bits 01, current 01
5797.376066216:imrelp.c       : imrelp.c: librelp: doing epoll_wait
5797.376072858:imrelp.c       : imrelp.c: librelp: done epoll_wait, nEvents:1
5797.376083197:imrelp.c       : imrelp.c: relpTcpRcv: read 97 bytes from sock 8
5797.376088809:imrelp.c       : imrelp.c: relp session read 97 octets, buf '1 open 86 relp_version=0
relp_software=librelp,1.2.15,http://librelp.adiscon.com
commands=syslog
'
5797.376099435:imrelp.c       : imrelp.c: relp engine is dispatching frame with command 'open'
5797.376104578:imrelp.c       : imrelp.c: in open command handler
5797.376113845:imrelp.c       : imrelp.c: processing client offer 'commands'
5797.376118141:imrelp.c       : imrelp.c: cmd syslog state in srv session: 3
5797.376122622:imrelp.c       : imrelp.c: SetEnableCmd in syslog cmd state: 3
5797.376127308:imrelp.c       : imrelp.c: SetEnableCmd out syslog cmd state: 4, iRet 0
5797.376131350:imrelp.c       : imrelp.c: processing client offer 'relp_software'
5797.376135538:imrelp.c       : imrelp.c: processing client offer 'relp_version'
5797.376154852:imrelp.c       : imrelp.c: relpTcpSend: sock 8, lenbuf 103, send returned 103 [errno 0]
5797.376160774:imrelp.c       : imrelp.c: in destructor: sendbuf 0x7f72a80244a0
5797.376204000:imrelp.c       : imrelp.c: librelp: epoll_set_events sock 8, target bits 01, current 01
5797.376210559:imrelp.c       : imrelp.c: librelp: doing epoll_wait
5797.379232131:imrelp.c       : imrelp.c: librelp: done epoll_wait, nEvents:1
5797.379277154:imrelp.c       : imrelp.c: relpTcpRcv: read 32768 bytes from sock 8
5797.379363561:imrelp.c       : imrelp.c: relp session read 32768 octets, buf '2 syslog 131072 <133>2018-04-10T10:36:33.317224+03:00 sender 1k ========<SNIP A LOT OF CHARS>========relp session 8 iRet 10010, tearing it down
5797.381108065:imrelp.c       : imrelp.c: librelp: delete sock 8 from epoll set
5797.381121090:imrelp.c       : imrelp.c: hint-frame to send: '0 serverclose 0
'
5797.381165568:imrelp.c       : imrelp.c: relpTcpSend: sock 8, lenbuf 16, send returned 16 [errno 0]
5797.381172268:imrelp.c       : imrelp.c: in destructor: sendbuf 0x7f72a80224e0
5797.381214677:imrelp.c       : imrelp.c: librelp: doing epoll_wait
```

Here I pressed ^C, 
```sh
^C5845.539622705:main thread    : rsyslogd.c: DoDie called.
5845.539646232:main thread    : rsyslogd.c: exiting on signal 2
5845.539652090:main thread    : rsyslogd.c: Terminating input threads...
5845.539661553:main thread    : ../threads.c: request term via SIGTTIN for input thread 'imrelp' 0x7f72ad974700
5845.539729852:imrelp.c       : imrelp.c: imrelp: termination requested via SIGTTIN - telling RELP engine
5845.539770987:imrelp.c       : imrelp.c: librelp: done epoll_wait, nEvents:-1
5845.539781185:imrelp.c       : imrelp.c: librelp: delete sock 5 from epoll set
5845.539795123:imrelp.c       : imrelp.c: librelp: delete sock 6 from epoll set
5845.539814454:imrelp.c       : ../threads.c: thrdStarter: usrThrdMain imrelp - 0x7f72ad974700 returned with iRet 0, exiting now.
5845.539882580:main thread    : ../threads.c: non-cancel input thread termination succeeded for thread imrelp 0x7f72ad974700
5845.539964122:imrelp.c       : debug.c: destructor for debug call stack 0x7f72a80008c0 called
5845.540179106:main thread    : iminternal.c: signaling new internal message via SIGTTOU: ' [origin software="rsyslogd" swVersion="8.34.0" x-pid="110" x-info="http://www.rsyslog.com"] exiting on signal 2.'
5845.540600870:main thread    : rsyslogd.c: Terminating main queue...
5845.540614486:main thread    : main Q: queue.c: shutdown: begin to destruct queue
5845.540620984:main thread    : main Q: queue.c: initiating worker thread shutdown sequence
5845.540626392:main thread    : main Q: queue.c: trying shutdown of regular workers
5845.540633713:main thread    : main Q: queue.c: regular queue workers shut down.
5845.540648423:main thread    : main Q: queue.c: checking to see if we need to cancel any worker threads of the primary queue
5845.540654849:main thread    : main Q: queue.c: worker threads terminated, remaining queue size log 0, phys 0.
5845.540684803:main thread    : main Q: queue.c: queue (type 0) will lose 0 messages, destroying...
5845.540725374:main thread    : rsyslogd.c: Terminating outputs...
5845.540732263:main thread    : ruleset.c: rulesetDestructAllActions
5845.540737271:main thread    : ruleset.c: RRRRRR: rsconfDestruct - queue shutdown
5845.540744297:main thread    : ruleset.c: shutting down queue workers for ruleset 0x7f72b6d3f810, name RSYSLOG_DefaultRuleset, queue (nil)
5845.540750558:main thread    : ruleset.c: shutting down queue workers for ruleset 0x7f72b6d6b880, name input, queue 0x7f72b6d6c1e0
5845.540755944:main thread    : input: queue.c: initiating worker thread shutdown sequence
5845.540761380:main thread    : input: queue.c: trying shutdown of regular workers
Segmentation fault (core dumped)
```

On different run I executed kill $pid
```sh
6148.506623747:main thread    : rsyslogd.c: DoDie called.
6148.506648944:main thread    : rsyslogd.c: exiting on signal 15
6148.506654951:main thread    : rsyslogd.c: Terminating input threads...
6148.506664376:main thread    : ../threads.c: request term via SIGTTIN for input thread 'imrelp' 0x7f1d3502d700
6148.506888178:imrelp.c       : imrelp.c: imrelp: termination requested via SIGTTIN - telling RELP engine
6148.506913795:imrelp.c       : imrelp.c: librelp: done epoll_wait, nEvents:-1
6148.506923028:imrelp.c       : imrelp.c: librelp: delete sock 5 from epoll set
6148.506934576:imrelp.c       : imrelp.c: librelp: delete sock 6 from epoll set
6148.506951330:imrelp.c       : ../threads.c: thrdStarter: usrThrdMain imrelp - 0x7f1d3502d700 returned with iRet 0, exiting now.
6148.507091759:imrelp.c       : debug.c: destructor for debug call stack 0x7f1d300008c0 called
6148.507314118:main thread    : ../threads.c: non-cancel input thread termination succeeded for thread imrelp 0x7f1d3502d700
6148.507350330:main thread    : iminternal.c: signaling new internal message via SIGTTOU: ' [origin software="rsyslogd" swVersion="8.34.0" x-pid="119" x-info="http://www.rsyslog.com"] exiting on signal 15.'
6148.507634337:main thread    : rsyslogd.c: Terminating main queue...
6148.507648656:main thread    : main Q: queue.c: shutdown: begin to destruct queue
6148.507655391:main thread    : main Q: queue.c: initiating worker thread shutdown sequence
6148.507660775:main thread    : main Q: queue.c: trying shutdown of regular workers
6148.507667752:main thread    : main Q: queue.c: regular queue workers shut down.
6148.507672772:main thread    : main Q: queue.c: checking to see if we need to cancel any worker threads of the primary queue
6148.507679512:main thread    : main Q: queue.c: worker threads terminated, remaining queue size log 0, phys 0.
6148.507700645:main thread    : main Q: queue.c: queue (type 0) will lose 0 messages, destroying...
6148.507737547:main thread    : rsyslogd.c: Terminating outputs...
6148.507744555:main thread    : ruleset.c: rulesetDestructAllActions
6148.507749760:main thread    : ruleset.c: RRRRRR: rsconfDestruct - queue shutdown
6148.507756811:main thread    : ruleset.c: shutting down queue workers for ruleset 0x7f1d3d57b810, name RSYSLOG_DefaultRuleset, queue (nil)
6148.507763175:main thread    : ruleset.c: shutting down queue workers for ruleset 0x7f1d3d5a7880, name input, queue 0x7f1d3d5a81e0
6148.507768824:main thread    : input: queue.c: initiating worker thread shutdown sequence
6148.507774217:main thread    : input: queue.c: trying shutdown of regular workers
Segmentation fault (core dumped)
```

### Environment
```sh
rsyslogd 8.34.0, compiled with:
        PLATFORM:                               x86_64-redhat-linux-gnu
        PLATFORM (lsb_release -d):
        FEATURE_REGEXP:                         Yes
        GSSAPI Kerberos 5 support:              Yes
        FEATURE_DEBUG (debug build, slow code): No
        32bit Atomic operations supported:      Yes
        64bit Atomic operations supported:      Yes
        memory allocator:                       system default
        Runtime Instrumentation (slow code):    No
        uuid support:                           Yes
        systemd support:                        Yes
        Number of Bits in RainerScript integers: 64

See http://www.rsyslog.com for more information.
```

```sh
Name        : rsyslog
Version     : 8.34.0
Release     : 2.el7.centos
Architecture: x86_64
Install Date: Tue Apr 10 10:28:57 2018
Group       : System Environment/Daemons
Size        : 2684739
License     : (GPLv3+ and ASL 2.0)
Signature   : RSA/SHA1, Wed Apr  4 14:46:22 2018, Key ID e0f233b3e00b8985
Source RPM  : rsyslog-8.34.0-2.el7.centos.src.rpm
Build Date  : Wed Apr  4 14:03:46 2018
Build Host  : localhost
Relocations : (not relocatable)
URL         : http://www.rsyslog.com/
Summary     : Enhanced system logging and kernel message trapping daemon
Description :
Rsyslog is an enhanced, multi-threaded syslog daemon. It supports MySQL,
syslog/TCP, RFC 3195, permitted sender lists, filtering on any message part,
and fine grain output format control. It is compatible with stock sysklogd
and can be used as a drop-in replacement. Rsyslog is simple to set up, with
advanced features suitable for enterprise-class, encryption-protected syslog
relay chains.
```

```sh
CentOS Linux release 7.4.1708 (Core)
Linux hostname 3.10.0-514.el7.x86_64 #1 SMP Tue Nov 22 16:42:41 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by StrongestNumber9** [comment]
Attached valgrind log for receiver
[valgrind-of-receiver.txt](https://github.com/rsyslog/rsyslog/files/1893612/valgrind-of-receiver.txt)

**Comment by rgerhards** [comment]
see also https://github.com/rsyslog/rsyslog/issues/2633

**Comment by rgerhards** [comment]
fix needs updated librelp 1.2.16 (to be released at latest with rsyslog 8.35.0)
as well as PR https://github.com/rsyslog/rsyslog/pull/2656

With #2656, the default is to truncate.

**Comment by rgerhards** [comment]
now merged, closed via #2689
